src/emqtt_bench.erl

```diff
@@ -645,25 +650,31 @@ loop(Parent, N, Client, PubSub, Opts) ->
 	end.
 
 put_publish_begin_time() ->
-    case get(last_publish_ts) of
-        undefined ->
-            NowT = erlang:monotonic_time(millisecond),
-            put(last_publish_ts, NowT);
-        _ ->
-            ok
-    end,
+    NowT = erlang:monotonic_time(millisecond),
+    put(publish_begin_ts, NowT),
     ok.
 
-next_publish(Opts) ->
-    inc_counter(pub),
-    BeginTime = get(last_publish_ts),
-    PubCnt = get_counter(pub),
-    Interval = proplists:get_value(interval_of_msg, Opts),
-    NextTime = BeginTime + PubCnt * Interval,
+get_publish_begin_time() ->
+    get(publish_begin_ts).
+
+%% @doc return new value
+bump_publish_attempt_counter() ->
+    NewCount = case get(success_publish_count) of
+                   undefined ->
+                       1;
+                   Val ->
+                       Val + 1
+               end,
+    _ = put(success_publish_count, NewCount),
+    NewCount.
+
+schedule_next_publish(Interval) ->
+    PubAttempted = bump_publish_attempt_counter(),
+    BeginTime = get_publish_begin_time(),
+    NextTime = BeginTime + PubAttempted * Interval,
```

Comment:
This seems to grow without bounds and eventually stop publishing, since `PubAttempted` only increases?

Reply:
It’s subtracted by the current ts when calculating the sleep interval

Reply:
good catch, I have moved the begin time `put` operation to before `loop`

Reply:
sorry, missed the comment above.
moved the initialization to publish/2